### PR TITLE
add produce blocks task

### PR DESCRIPTION
**Pull request by azban**
here's the task that will consume the stream of `ConsensusBlockBody` that txflow creates. 

i extracted out a `BeaconBlockProducer` from the client.. which ended up requiring a `Runtime` so i had to refactor `StateDbViewer` a bit.

### CODE REVIEW

File: node/beacon-chain-handler/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+use primitives::types::ConsensusBlockBody;
+use primitives::types::SignedTransaction;
+use std::sync::Arc;
+use storage::StateDb;
+
+pub trait ConsensusHandler<B: Block, P>: Send + Sync {
+    fn produce_block(&self, body: ConsensusBlockBody<P>) -> B;
+}
+
+pub struct BeaconBlockProducer {
```
Comment (ilblackdragon) [inline node/beacon-chain-handler/src/lib.rs]: seems like we split Client into 3 objects (BeaconBlockProducer, StateDbViewer and whatever future object that will import blocks?) which have exactly the same content as `Client` had but are across different cargos?

File: node/beacon-chain-handler/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            signer,
+            state_db,
+        }
+    }
+}
+
+pub type BeaconChainConsensusBlockBody = ConsensusBlockBody<Vec<SignedTransaction>>;
+
+impl ConsensusHandler<BeaconBlock, Vec<SignedTransaction>> for BeaconBlockProducer {
+    fn produce_block(&self, body: BeaconChainConsensusBlockBody) -> BeaconBlock {
```
Comment (ilblackdragon) [inline node/beacon-chain-handler/src/lib.rs]: if we are trying to build real interface - we should remove signing from here, and it should just return back to txflow what they going to be signing on (hash of the block).
The constructed block with state should be put in some cache, until it's signed and only then added to the chain.

Comment (nearmax) [reply node/beacon-chain-handler/src/lib.rs]: If we implement future interface for the signer then this method can return future too and there is no need to store the unsigned block somewhere (technically, the unsigned block will temporarily live inside the stack of the produce_block function while its future is pending)

File: node/beacon-chain-handler/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            runtime,
+            signer,
+            state_db,
+        }
+    }
+}
+
+pub type BeaconChainConsensusBlockBody = ConsensusBlockBody<Vec<SignedTransaction>>;
+
+impl ConsensusHandler<BeaconBlock, Vec<SignedTransaction>> for BeaconBlockProducer {
```
Comment (nearmax) [inline node/beacon-chain-handler/src/lib.rs]: It is good enough for now, later we might get rid of the trait or the struct.

File: node/beacon-chain-handler/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        self.state_db.commit(&mut apply_result.transaction).ok();
+        let mut block = BeaconBlock::new(
+            last_block.header().index() + 1,
+            last_block.hash(),
+            apply_result.root,
+            filtered_transactions,
+        );
+        block.sign(&self.signer);
+        self.beacon_chain.insert_block(block.clone());
+        block
```
Comment (nearmax) [inline node/beacon-chain-handler/src/lib.rs]: Why do we need to return it?

File: node/beacon-chain-handler/src/lib.rs (hunk 1 above)
Comment (azban) [reply node/beacon-chain-handler/src/lib.rs]: they don't need to be split up between crates but i wanted to just start extracting things from client. i agree we should do some crate consolidation in the future. it seems like we are maybe abusing crates in general and should lean towards modules in some places.

as to why i am splitting them up: they don't have the same content. `BlockImporter` may be implemented with the producer.. not sure yet.

```
pub struct StateDbViewer {
    beacon_chain: Arc<BlockChain<BeaconBlock>>,
    state_db: Arc<StateDb>,
}
```

```
pub struct BeaconBlockProducer {
    beacon_chain: Arc<BlockChain<BeaconBlock>>,
    runtime: Runtime,
    signer: Arc<Signer>,
    state_db: Arc<StateDb>,
}
```

part of the motivation behind separating the client is to eliminate a god object being relied upon everywhere. another part is to eventually get to a point where we have interfaces that are copyable without worrying about write locks across threads. i think we are stuck on blockchain's and state db's, but have some ideas for how to solve elegantly without locking in the future (like providing eventual consistency on reads from the `StateDbViewer`). @nearmax may be able to comment more.

File: node/beacon-chain-handler/src/lib.rs (hunk 4 above)
Comment (azban) [reply node/beacon-chain-handler/src/lib.rs]: we don't, i can remove this

**Pull request merged**